File: src/p0002/erlang/solution.erl
-spec add_two_numbers(A :: #list_node{} | null, B :: #list_node{} | null) -> #list_node{} | null.
add_two_numbers(A, B) ->
    add_two_numbers(A, B, 0).

add_two_numbers(null, null, 0) ->
    null;
add_two_numbers(A, B, Carry) ->
    {ValA, NextA} = extract(A),
    {ValB, NextB} = extract(B),
    C = Carry + ValA + ValB,
    #list_node{val = C rem 10, next = add_two_numbers(NextA, NextB, C div 10)}.

extract(#list_node{val = Val, next = Next}) -> {Val, Next};
extract(_) -> {0, null}.

File: src/p0002/elixir/solution.ex
defmodule Solution do
  @spec add_two_numbers(a :: ListNode.t | nil, b :: ListNode.t | nil) :: ListNode.t | nil
  def add_two_numbers(a, b), do: add_two_numbers(a, b, 0)

  def add_two_numbers(nil, nil, 0), do: nil
  def add_two_numbers(a, b, carry) do
    {val_a, next_a} = extract(a)
    {val_b, next_b} = extract(b)
    carry = carry + val_a + val_b
    %ListNode{val: rem(carry, 10), next: add_two_numbers(next_a, next_b, div(carry, 10))}
  end

  def extract(%ListNode{val: val, next: next}), do: {val, next}
  def extract(_), do: {0, nil}
end
